### PR TITLE
Add AdapterMessageConverter trait and a payload serializer

### LINKED ISSUE #56: Provide helper trait for adapters to centralize message conversion logic
- [x] Let the trait handle the new adapter options `message_factory` and `message_converter`
- [x] Provide trait methods: `messageToArray`, `messageFromArray`
- [x] May also centralize the payload serializer handling but keep in mind that the mongo db adapter does not need it, so it should be optional

### PR DESCRIPTION
**Pull request by codeliner**
- Closes #56

### CODE REVIEW

**Comment by codeliner** [comment]
ping @prolic and the next one ;) This will simplify the adjustment of the adapters.

File: src/Adapter/PayloadSerializer.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    public function serializePayload(array $payload);
+
+    /**
+     * @param string $payloadStr
+     * @return array
+     */
+    public function unserializePayload($payloadStr);
+}
+
```
Comment (prolic) [inline src/Adapter/PayloadSerializer.php]: missing new line at end of file, again ;)

File: src/Adapter/Exception/RuntimeException.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Class RuntimeException
+ *
+ * @package Prooph\EventStore\Adapter\Exception
+ * @author Alexander Miertsch <alexander.miertsch.extern@sixt.com>
+ */
+final class RuntimeException extends \RuntimeException implements AdapterException
+{
+}
+
```
Comment (prolic) [inline src/Adapter/Exception/RuntimeException.php]: missing new line at end of file, again ;)

File: src/Adapter/AdapterMessageConverter.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            '%s::%s: not possible. No %s set. May forgot to invoke %s::%s?',
+            get_class($this),
+            $inMethod,
+            $dependency,
+            AdapterMessageConverter::class,
+            'initMessageConversion'
+        ));
+    }
+}
+
```
Comment (prolic) [inline src/Adapter/AdapterMessageConverter.php]: missing new line at end of file

File: src/Adapter/PayloadSerializer/JsonPayloadSerializer.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * @param string $payloadStr
+     * @return array
+     */
+    public function unserializePayload($payloadStr)
+    {
+        return json_decode($payloadStr, true);
+    }
+}
+
```
Comment (prolic) [inline src/Adapter/PayloadSerializer/JsonPayloadSerializer.php]: missing new line at end of file

File: tests/Adapter/AdapterMessageConverterTest.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+        return [
+            ["createMessageFromArray", ["test_message", []]],
+            ["convertMessageToArray", [$message->reveal()]],
+            ["serializePayload", [[]]],
+            ["unserializePayload", ["payload"]],
+        ];
+    }
+}
+
```
Comment (prolic) [inline tests/Adapter/AdapterMessageConverterTest.php]: missing new line at end of file

File: tests/Adapter/PayloadSerializer/JsonPayloadSerializerTest.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                ['bool_true' => true, 'bool_false' => false],
+                ['int' => 1234],
+                ['float' => 10.2],
+                ['null' => null],
+                ['array' => ['nested' => ['data' => 'structure']]],
+            ]
+        ];
+    }
+}
+
```
Comment (prolic) [inline tests/Adapter/PayloadSerializer/JsonPayloadSerializerTest.php]: missing new line at end of file

File: tests/Mock/AdapterMessageConverterMock.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        return $this->serializePayload($payload);
+    }
+
+    public function proxyToUnserializePayload($payloadStr)
+    {
+        return $this->unserializePayload($payloadStr);
+    }
+}
+
```
Comment (prolic) [inline tests/Mock/AdapterMessageConverterMock.php]: missing new line at end of file

File: src/Adapter/AdapterMessageConverter.php [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+/**
+ * Trait AdapterMessageConverter
+ *
+ * This trait adds message conversion capabilities to any event store adapter.
+ *
+ * @package Prooph\EventStore\Adapter
+ * @author Alexander Miertsch <alexander.miertsch.extern@sixt.com>
+ */
+trait AdapterMessageConverter
```
Comment (prolic) [inline src/Adapter/AdapterMessageConverter.php]: I would propose renaming this class to AdapterMessageConverterTrait.

File: src/Adapter/AdapterMessageConverter.php [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+use Prooph\Common\Messaging\Message;
+use Prooph\Common\Messaging\MessageConverter;
+use Prooph\Common\Messaging\MessageFactory;
+use Prooph\EventStore\Adapter\Exception\ConfigurationException;
+use Prooph\EventStore\Adapter\Exception\RuntimeException;
+
+/**
+ * Trait AdapterMessageConverter
+ *
+ * This trait adds message conversion capabilities to any event store adapter.
```
Comment (prolic) [inline src/Adapter/AdapterMessageConverter.php]: Perhabs hint, that not all adapters need to serialize the payload (f.e. mongo db).

File: src/Adapter/PayloadSerializer.php (hunk 1 above)
Comment (codeliner) [inline src/Adapter/PayloadSerializer.php]: hhhm, what exactly is missing? And how can I change it? Thought before you mean that I should add an empty line after the closing bracket but you mean something different, right?
My phpstorm is set up to use \n as line ending. What am I doing wrong?

File: src/Adapter/AdapterMessageConverter.php (hunk 8 above)
Comment (codeliner) [inline src/Adapter/AdapterMessageConverter.php]: yea I first named it this way but I stopped putting Interface behind each interface. To be consistent with naming I think it is the same for traits. IMO the name should only represent the main purpose of the class, trait, interface, whatever. That it is a trait is already covered by the definition.

Comment (prolic) [inline src/Adapter/AdapterMessageConverter.php]: Ok
Am 14.08.2015 23:04 schrieb "Alexander Miertsch" <notifications@github.com>:

File: src/Adapter/PayloadSerializer.php (hunk 1 above)
Comment (prolic) [inline src/Adapter/PayloadSerializer.php]: There is a space in the last line. It should be an empty line.

Comment (prolic) [inline src/Adapter/PayloadSerializer.php]: :-)

**Comment by codeliner** [comment]
@prolic sorry for the missing line endings. I guess my phpstorm class template is not working correctly. Maybe I should update my phpstorm installation ... Will keep an eye on it in the future. promised :smile: 

I added a hint like suggested. 
Regarding the naming please see my comment. What are the pros for a `Trait`  suffix?

**Comment by codeliner** [comment]
oh, forget my question. Missed your comment ...

**Comment by prolic** [comment]
No problem. But I like to mention, that in the zf2 team, we made the same decision during beta phase first. We remove all interface suffixes from classes (by poll of core devs), but then ended up with a directories full of php files, and you had to open every one (or have a good ide with hints) to know, whether this is an interface or an actual implementation.
In the end, we made a new poll and reverted this.

If you look up in the zf2 mailing list, you can find the related discussion.

**Comment by prolic** [comment]
Additional note:

I also would not use Interface suffixes within a domain model, as this is about a ubiquitous language. You don't get the domain experts to talk about a UserCollectionInterface, they only talk about a UserCollection.

But within a framework (and I would treat the event-store as framework, not a domain) I prefere Interfaces suffixes.

That said, in a project (like proophessor-do) you'll have an EventStoreAdapterInterface (interface within framework) and a UserCollection (interface within the domain). But I can live with it, if you decide otherwise.

**Comment by codeliner** [comment]
@prolic yeah I remember a very long discussion in the DDDinPHP group about the same topic. 
Mathias wrote a very good article about it in 2013: http://verraes.net/2013/09/sensible-interfaces/
I can follow your argument with DDD vs. framework. But on the other hand I think the cons explained by Mathias apply also to framework naming, because the way of thinking about interfaces and implementations is different. Last but not least, the prooph components provide DDD infrastructure and should therefor follow a DDD like naming. Broadway for example will also remove the interface suffix if not already done.

**Comment by codeliner** [comment]
@prolic and you'll also get better ide support by the way. Try to type hint in a Psr-7 middleware __invoke method for `RequestInterface`. 

image

I hate it! If the interface would only be `Request` I could press enter and continue writing my code.

**Comment by prolic** [comment]
Yeah with phpstorm
Some devs (like Matthew Weier O'Phinney) use vim! ;)

Was just a hint, it's really not important to me. I can live with it either way.

**Comment by codeliner** [comment]
yeah I also know some devs working with vim. crazy people :-). vim is great but I don't want to spent 10 years only to learn all the short cuts. Once refactored some code with phpstorm (what you have to do very often in a DDD project) and you will never ever use something other.
However, thanks for the discussion. Sometimes these topics need to be discussed. My point of view can be wrong and a second opinion is always a good thing. So thanks again for your reviews and taking the time to discuss these things with me. So time to merge now.

**Comment by codeliner** [comment]
@prolic hehe, just read the comments of Mathias' article again. Ocramius joined the discussion and wrote about the ZF2 decision you mentioned. That was the discussion I had in mind earlier. Was not on the DDDinPHP group. There we had a discussion about the `final` keyword and `private` vs `protected`.

**Pull request merged**